src/faction_camp.cpp

```diff
@@ -0,0 +1,3349 @@
+                if( comp != nullptr ) {
+                    popup( _( "%s returns from your garage..." ), comp->name.c_str() );
+                    camp_companion_return( *comp );
+                }
+                break;
+            }
+        }
+    }
+
+    g->draw_ter();
+    wrefresh( g->w_terrain );
+
+    return true;
+}
+
+// camp faction companion mission start functions
+void talk_function::start_camp_upgrade( npc &p, const std::string &bldg )
+{
+    std::vector<std::shared_ptr<npc>> npc_list = companion_list( p, "_faction_upgrade_camp" );
+    if( !npc_list.empty() ) {
+        popup( _( "You already have a companion upgrading the camp." ) );
+        return;
+    }
+    const recipe *making = &recipe_id( bldg ).obj();
+    //Stop upgrade if you don't have materials
+    inventory total_inv = g->u.crafting_inventory();
+    if( making->requirements().can_make_with_inventory( total_inv, 1 ) ) {
+        int need_food = time_to_food( time_duration::from_turns( making->time / 100 ) );
+        if( camp_food_supply() < need_food && bldg != "faction_base_camp_1" ) {
+            popup( _( "You don't have enough food stored to feed your companion." ) );
+            return;
+        }
+        npc *comp = individual_mission( p, _( "begins to upgrade the camp..." ), "_faction_upgrade_camp",
+                                        false, {},
+                                        making->skill_used.obj().ident().c_str(), making->difficulty );
+        if( comp != nullptr ) {
+            comp->companion_mission_time_ret = calendar::turn + time_duration::from_turns(
+                                                   making->time / 100 ) ;
+            g->u.consume_components_for_craft( making, 1, true );
+            g->u.invalidate_crafting_inventory();
+        }
+    } else {
+        popup( _( "You don't have the materials for the upgrade." ) );
+    }
+}
+
+void talk_function::start_cut_logs( npc &p )
+{
+    std::vector<std::shared_ptr<npc>> npc_list = companion_list( p, "_faction_camp_cut_log" );
+    if( !npc_list.empty() ) {
+        popup( _( "There are too many companions working on this mission!" ) );
+        return;
+    }
+    std::vector<std::string> log_sources = { "forest", "forest_thick", "forest_water" };
+    popup( _( "Forests and swamps are the only valid cutting locations." ) );
+    const tripoint omt_pos = p.global_omt_location();
+    tripoint forest = om_target_tile( omt_pos, 1, 50, log_sources );
+    if( forest != tripoint( -999, -999, -999 ) ) {
+        int tree_est = om_harvest_trees( p, forest, .50, false, false );
+        int tree_young_est = om_harvest_ter( p, forest, ter_id( "t_tree_young" ), .50, false );
+        int dist = rl_dist( forest.x, forest.y, omt_pos.x, omt_pos.y );
+        //Very roughly what the player does + 6 hours for prep, clean up, breaks
+        time_duration chop_time = 6_hours + ( 1_hours * tree_est ) + ( 7_minutes * tree_young_est );
+        //Generous to 
```

Comment:
This id type should be returned. The caller can use it directly and it documents what the function actually returns. If the caller absolutely needs a string, they can do the conversion on their own.

Reply:
Although I agree with you in principle, the code works now with the string references and my first attempt at fixing them caused CDDA to crash on program start. Since my goal in this refactor is to just the code to the start where I can work on it without refactoring it to minimal standards, I will leave this as is.